go/src/pluginmgr/pluginmgr_common/pluginmgr_helper.go

```diff
@@ -0,0 +1,505 @@
+	"fmt"
+	"go/src/lib/lomcommon"
+	"go/src/lib/lomipc"
+	"go/src/plugins/plugins_common"
+	"go/src/plugins/plugins_files"
+	"log/syslog"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+)
+
+/*****************************************************************************************/
+/* Plugin manager */
+/*****************************************************************************************/
+
+// Plugin Manager global variables
+var (
+	goroutinetracker *lomcommon.GoroutineTracker = nil
+	ProcID           string                      = ""
+	pluginMgr        *PluginManager              = nil
+)
+
+// TODO: Goutham : Move this to global_conf.json
+// Constants for plugin manager with default values
+const (
+	GOLIB_TIMEOUT_DEFAULT                         = 0 * time.Millisecond /* Default GoLIB API timeouts */
+	PLUGIN_INIT_PERIODIC_FALLBACK_TIMEOUT_DEFAULT = 3600 * time.Second   /* Default Periodic logging long timeout */
+	PLUGIN_INIT_PERIODIC_TIMEOUT_DEFAULT          = 300 * time.Second    /* Default periodic logging short timeout */
+	PLUGIN_LOADING_TIMEOUT_DEFAULT                = 30 * time.Second
+	APP_NAME_DEAULT                               = "plgMgr"
+)
+
+// Plugin Manager interface to be implemented by plugin manager
+type IPluginManager interface {
+	run(chan struct{}) error
+	shutdown() error
+	addPlugin(pluginName string, pluginVersion string) error
+	registerActionWithEngine(pluginName string) error
+	deRegisterActionWithEngine(pluginName string) error
+	addPeriodicLogWithTimeouts(ID string, message string, shortTimeout time.Duration, longTimeout time.Duration, context string) chan bool
+	loadPlugin(pluginName string, pluginVersion string) error
+	getPlugin(plgname string) (plugins_common.Plugin, bool)
+	getPluginMetadata(plgname string) (plugins_common.IPluginMetadata, bool)
+}
+
+// Interface  for lomipc.ClientTx
+type IClientTx interface {
+	RegisterClient(client string) error
+	DeregisterClient() error
+	RegisterAction(action string) error
+	DeregisterAction(action string) error
+	RecvServerRequest() (*lomipc.ServerRequestData, error)
+	SendServerResponse(res *lomipc.MsgSendServerResponse) error
+	NotifyHeartbeat(action string, tstamp int64) error
+}
+
+type PluginManager struct {
+	clientTx IClientTx
+	plugins  map[string]plugins_common.Plugin /* map : pluginname -> Plugin struct Object(at go/src/plugins/plugins_common)
+	e.g. "linkflapdetection" -->  {linkFlapDetection object} */
+	pluginMetadata map[string]plugins_common.IPluginMetadata /* map : pluginname -> PluginMetadata struct Object(at go/src/plugins/plugins_common) */
+	stopch         chan struct{}                             /* channel used to stop listening on golib for server events */
+	//TODO: Goutham : replace this with global stop channel
+}
+
+// InitPluginManager : Initialize plugin manager & register with server
+func NewPluginManager(clientTx IClientTx) (*PluginManager, error) {
+	if pluginMgr != nil {
+		return pluginMgr, errors.New("client already registered")
```

Comment:
Panic or
return { nil & error }

-- Find & fix -- if this happens

Reply:
added